src/mlpack/methods/pca/decomposition_policies/stochastic_method.hpp

```diff
@@ -0,0 +1,323 @@
+   * exact SVD method.
+   *
+   * @param data Data matrix.
+   * @param centeredData Centered data matrix.
+   * @param transformedData Matrix to put results of PCA into.
+   * @param eigVal Vector to put eigenvalues into.
+   * @param eigvec Matrix to put eigenvectors (loadings) into.
+   * @param rank Rank of the decomposition.
+   */
+  void Apply(const arma::mat& data,
+             const arma::mat& centeredData,
+             arma::mat& transformedData,
+             arma::vec& eigVal,
+             arma::mat& eigvec,
+             const size_t rank)
+  {
+    PCAFunction f(centeredData);
+
+    // Adjust the step size if necessary.
+    SetStepSize(optimizer);
+
+    eigvec = arma::randn(centeredData.n_rows, rank);
+    optimizer.Optimize(f, eigvec);
+
+    // Project the samples to the principals.
+    transformedData = arma::trans(eigvec) * centeredData;
+
+    // Approximate eigenvalues and eigenvectors using Rayleigh–Ritz method.
+    arma::mat u, v;
+    arma::svd_econ(u, eigVal, v, transformedData);
+
+    // Now we must square the singular values to get the eigenvalues.
+    // In addition we must divide by the number of points, because the
+    // covariance matrix is X * X' / (N - 1).
+    eigVal %= eigVal / (data.n_cols - 1);
+  }
+
+  //! Get the optimizer.
+  OptimizerType Optimizer() const { return optimizer; }
+  //! Modify the optimizer.
+  OptimizerType& Optimizer() { return optimizer; }
+
+ private:
+  //! Adjust the step size if required.
+  template<typename T>
+  inline typename std::enable_if<
+      HasStepSize<T, double&(T::*)()>::value, void>::type
+  SetStepSize(T& optimizer)
+  {
+    if (optimizer.StepSize() > 0)
+    {
+      Log::Warn << "StepSize(): invalid value (> 0), automatically take the "
+          << "negative direction." << std::endl;
+      optimizer.StepSize() *= -1;
+    }
+  }
+
+  //! Adjust the step size if required.
+  template<typename T>
+  inline typename std::enable_if<
+      !HasStepSize<T, double&(T::*)()>::value, void>::type
+  SetStepSize(T& /* optimizer */)
+  {
+    /* Nothing to do here */
+  }
+
+  OptimizerType optimizer;
+
+  /**
+   * The principal component analysis (PCA) function for the
+   * Principal component analysis (PCA). This is used by various mlpack
+   * optimizers to be used for principal component analysis.
+   */
+  class PCAFunction
+  {
+   public:
+    /**
+     * Construct The principal component analysis (PCA) function with the
+     * given data.
+     *
+     * @param data Data matrix.
+     */
+    PCAFunction(const arma::mat& data) : data(data), pl(0)
+    {
+      visitationOrder = arma::linspace<arma::Row<size_t> >(0,
+          data.n_cols - 1, data.n_cols);
+    }
+
+    /**
+     * Shuffle the order of function visitation. This may be called by the
+     * optimizer.
+     */
+    void Shuffle()
+    {
+      visitationOrder = arma::shuffle(
+          arma::linspace<arma::Row<size_t> >(0, data.n_cols - 1, data.n_cols));
```

Comment:
It might give a nice speedup to shuffle the entire data matrix instead of keeping a `visitationOrder`.